### PR TITLE
added huggingface cli tests

### LINKED ISSUE #727: Add Hugging Face Hub integration
**Is your feature request related to a problem? Please describe.**
Sharing pretrained GaNDLF models with the ML community is difficult but important for collaboration and open science.

**Describe the solution you'd like**
Hugging Face Hub is a great option for sharing models, so a native integration with the HF Hub would make sharing and loading models easy.

**Describe alternatives you've considered**
N/A

**Additional context**
Have discussed with @sarthakpati :)

### PR DESCRIPTION
**Pull request by pranayasinghcsmpl**
Fixes #727 

## Proposed Changes

- Added test cases in testing/entrypoints

## Checklist

- [ ] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [ ] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [x] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [ ] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [x] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [x] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [x] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].

### CODE REVIEW

**Comment by sarthakpati** [comment]
I am assuming this PR makes #880 redundant.

**Comment by sarthakpati** [comment]
This needs fairly robust documentation regarding usage, including references to the various HF commands used internally [[ref](https://huggingface.co/docs/huggingface_hub/guides/overview)]. I would recommend creating a separate subsection under the [deployment header of usage](https://github.com/pranayasinghcsmpl/GaNDLF/blob/hf_tests/docs/usage.md#deployment) called `Deplying Models to HuggingFace Hub` and then adding details. It would be **awesome** if you could showcase an example in the form of a tutorial as well!

**Comment by pranayasinghcsmpl** [comment]
Sure, I'll get started on that.

**Comment by pranayasinghcsmpl** [comment]
> I am assuming this PR makes #880 redundant.

yeah, it does

File: docs/usage.md [hunk 1]
```diff
@@ -510,3 +510,522 @@ This can be replicated for ROCm for AMD , by following the [instructions to set
...
+```bash
+# Or using an environment variable
+>>> gandlf hf login --token $HUGGINGFACE_TOKEN --add-to-git-credential
+Token is valid (permission: write).
+Your token has been saved in your configured git credential helpers (store).
+Your token has been saved to /home/wauplin/.cache/huggingface/token
+Login successful
+```
+
+For more details about authentication, check out [this section](../quick-start#authentication).
```
Comment (sarthakpati) [inline docs/usage.md:557]: There seem to be links starting with `../`, which I think are not going to be present since they are referring to HF documentation. Please link to the proper HTML page from HF for such links.

For example, `../quick-start#authentication` should be https://huggingface.co/docs/huggingface_hub/v0.23.4/en/quick-start#authentication

File: docs/usage.md [hunk 2]
```diff
@@ -510,3 +510,522 @@ This can be replicated for ROCm for AMD , by following the [instructions to set
...
+orgs:  huggingface,eu-test,OAuthTesters,hf-accelerate,HFSmolCluster
+```
+
+If you are not logged in, an error message will be printed.
+
+### huggingface-cli logout
+
+This commands logs you out. In practice, it will delete the token saved on your machine.
+
+This command will not log you out if you are logged in using the `HF_TOKEN` environment variable (see [reference](../package_reference/environment_variables#hftoken)). If that is the case, you must unset the environment variable in your machine configuration.
```
Comment (sarthakpati) [inline docs/usage.md:575]: Refer to my previous comment about having `../` links.

File: docs/usage.md [hunk 3]
```diff
@@ -510,3 +510,522 @@ This can be replicated for ROCm for AMD , by following the [instructions to set
...
+
+### Download to a local folder
+
+The recommended (and default) way to download files from the Hub is to use the cache-system. However, in some cases you want to download files and move them to a specific folder. This is useful to get a workflow closer to what git commands offer. You can do that using the `--local-dir` option.
+
+A `./huggingface/` folder is created at the root of your local directory containing metadata about the downloaded files. This prevents re-downloading files if they're already up-to-date. If the metadata has changed, then the new file version is downloaded. This makes the `local-dir` optimized for pulling only the latest changes.
+
+<Tip>
+
+For more details on how downloading to a local file works, check out the [download](./download.md#download-files-to-a-local-folder) guide.
```
Comment (sarthakpati) [inline docs/usage.md:666]: Refer to my previous comment about having `./` or `../` links.

File: docs/usage.md [hunk 4]
```diff
@@ -510,3 +510,522 @@ This can be replicated for ROCm for AMD , by following the [instructions to set
 GaNDLF, and GaNDLF-created models, may be distributed as an [MLCube](https://mlcommons.github.io/mlcube/). This involves distributing an `mlcube.yaml` file. That file can be specified when using the [MLCube runners](https://mlcommons.github.io/mlcube/runners/). The runner will perform many aspects of configuring your container for you. Currently, only the `mlcube_docker` runner is supported. 
 
 See the [MLCube documentation](https://mlcommons.github.io/mlcube/) for more details.
+
+## Hugging Face Hub CLI
```
Comment (sarthakpati) [inline docs/usage.md:514]: Minor edit to section heading
```suggestion
## GaNDLF Integration with Hugging Face Hub
```

File: docs/usage.md [hunk 5]
```diff
@@ -510,3 +510,522 @@ This can be replicated for ROCm for AMD , by following the [instructions to set
...
+
+This tool allows you to interact with the Hugging Face Hub directly from a terminal. For example, you can login to your account, create a repository, upload and download files, etc. It also comes with handy features to configure your machine or manage your cache.
```
Comment (sarthakpati) [inline docs/usage.md:516]: The level of detail you have provided is great, but it seems like you have pulled information verbatim from [HF site](https://huggingface.co/docs/huggingface_hub/v0.23.4/en/guides/cli). 

If the usage of `gandlf hf` is identical to `huggingface-cli`, please say so and put a link to the HF CLI usage in this section.

**Review by sarthakpati — CHANGES_REQUESTED** [review]

File: docs/usage.md (hunk 5 above)
Comment (sarthakpati) [reply docs/usage.md:516]: > The level of detail you have provided is great, but it seems like you have pulled information verbatim from [HF site](https://huggingface.co/docs/huggingface_hub/v0.23.4/en/guides/cli).
> 
> If the usage of `gandlf hf` is identical to `huggingface-cli`, please say so and put a link to the HF CLI usage in this section.

This comment supersedes the others, so I have resolved them.

**Comment by sarthakpati** [comment]
Hi @Wauplin @NielsRogge, would you mind taking a look at this PR to see it makes sense from the HF perspective?

**Comment by Wauplin** [comment]
Hi @sarthakpati, thanks for the ping. I've reviewed this PR + had a look to https://github.com/mlcommons/GaNDLF/issues/727 https://github.com/mlcommons/GaNDLF/pull/880 https://github.com/mlcommons/GaNDLF/pull/878. To be honest, I'm not sure to understand the added value of wrapping `huggingface-cli`. For what I understand, users will type `gandlf hf download ...` instead of `huggingface-cli download ...` but then the behavior will be exactly the same? It that's the case, I don't think it's worth adding it. The cost of maintenance (e.g. adding new commands when `huggingface_hub` gets updated) seems to outweigh the benefits in my opinion. If `gandlf` usage is purely CLI-based, I understand the need for a CLI. In that case, I would either:
- document `huggingface-cli upload` and `huggingface-cli download` in the documentation, without wrapping them
- or, have `gandlf hf download` and `gandlf hf upload` but only if additional features are added to it. For example default values from model ids depending on the GaNDLF model to load. Or make sure in `gandlf hf upload` that a model card is created for the trained model which would contains some information about the training + have some metadata. Metadata is important for discoverability on the Hub, say for example to filter all models with `gandlf` as library_name: https://huggingface.co/models?library=gandlf.


Overall, I'm new to the GaNDLF codebase and don't want to assume how users are using it. But if you have some script based usage (e.g. users importing GaNDLF classes in their code), then I think that a tight integration as suggested in https://github.com/mlcommons/GaNDLF/pull/878 by @nielsrogge makes more sense. Being able to load / save models on the HF Hub directly is usually a much appreciated feature.

**Comment by sarthakpati** [comment]
Hi @Wauplin,

Thank you for your feedback. I think the second option you suggest makes a lot of sense, since we definitely need to have added metadata. At the very least, we need to have the version of GaNDLF the model was trained on and (if applicable) the accompanying git hash. And having `gandlf as `library_name` would be perfect. 

@pranayasinghcsmpl - does this make sense?

**Comment by Wauplin** [comment]
Agree!

Btw, PR from @nielsrogge https://github.com/mlcommons/GaNDLF/pull/878/files would help doing that. In the Mixin, you can define metadata that will be added to the model card. Here is an integration guide about it: https://huggingface.co/docs/huggingface_hub/v0.24.0.rc0/en/guides/integrations#model-card. Once that's integrated, the `gandlf upload` and `gandlf download` command will simply be wrappers around the `from._pretrained` and `.push_to_hub` methods!

**Comment by pranayasinghcsmpl** [comment]
Hi @Wauplin, the GaNDLF cli exports other files apart from just the model like logs, csvs, onnx files etc. The gandlf cli has various commands to perform different stages of model development. So if an user uses `gandlf run` to train a model they can then use `gandlf collect-stats` to perform model evaluation which will generate more files containing plots & stats. Also to run inference on a gandlf model the user requires the config file that was used during training and all the files exported after training the model.

Therefore, it is important to be able to upload a lot of different files that are generated by different GaNDLF commands.
Along with that the user might want to include the training data that was used.

If we can upload all these files using the Mixin, we can just proceed with your suggestions.

**Comment by sarthakpati** [comment]
@pranayasinghcsmpl - I am unsure if uploading all the files to HF is required. I would imaging the model files (`*.pth.tar`, `*.onnx`, `*.xml`, `*.bin`) would be the most important, along with the model configuration itself. These combined with the metadata needed for HF hub to be able to parse the uploads as being `gandlf` centric (using the `library_name` key) should suffice.

**Comment by Wauplin** [comment]
Thanks for the broader context @pranayasinghcsmpl ! Given all the artifacts, it's indeed a bit more complex and the mixin won't solve it all. Then the best way to make a `gandlf hf upload` command is to use [`upload_folder`](https://huggingface.co/docs/huggingface_hub/guides/upload#upload-a-folder) on the folder with all the artifacts. Before that, you can generate a modelcard locally (with `library_name`) using the [ModelCard class](https://huggingface.co/docs/huggingface_hub/guides/model-cards). You can define a set of rules `allow_patterns`/`ignore_patterns` to tell which files you want to select. Same for `gandlf hf download`, using [`snapshot_download`](https://huggingface.co/docs/huggingface_hub/guides/download#download-an-entire-repository) with some `allow_patterns`/`ignore_patterns` would suffice. 

Let me know if you have any questions :hugs:

**Comment by pranayasinghcsmpl** [comment]
Hi, @Wauplin Thanks for the suggestions. We'll work on it & we'll let you know if we face any issues.

**Comment by pranayasinghcsmpl** [comment]
@Wauplin, is there any predefined way to set the version of the model in a model repo.

**Comment by Wauplin** [comment]
@pranayasinghcsmpl not really no. Some labs would do 1 version of the model == 1 repo on the Hub and use a convention in the repo name (e.g. "3" in  `meta-llama/Meta-Llama-3-8B`). the advantage is to differentiate usage, discussions, analytics, etc. between model versions. Another possibility is to upload different versions to the same repo and use git tags to version the model. You can use [`create_tag`](https://huggingface.co/docs/huggingface_hub/main/en/package_reference/hf_api#huggingface_hub.HfApi.create_tag) for that. The advantage is to keep everything in the same place, especially keeping the model card and discussions. The second solution is more suitable if every time you have a new version, you expect users to switch to this one instead of having usage on both.

**Comment by pranayasinghcsmpl** [comment]
@sarthakpati - Maybe we could save a copy of the config file in the model folder during training. There might be a better way to solve this. let me look into it.

**Comment by sarthakpati** [comment]
> @sarthakpati, could we maybe convert the parameters.pkl into a yaml format. Because it seems to have a lot of values that are in the config file & some that aren't.

Sure, but please add the yaml file as an addition the pkl (for backward compatibility reasons).

**Comment by pranayasinghcsmpl** [comment]
ok, sure

**Comment by sarthakpati** [comment]
Please change the base of this PR to master.

**Comment by sarthakpati** [comment]
Please change the base of this PR to master.

**Comment by pranayasinghcsmpl** [comment]
Sure, I'll do it in my next commit.
